### ISSUE TITLE
Cannot run simple example with either tsc or node

### ISSUE THREAD

**Issue by quanghoc** [jupyter/jupyter-js-services#99]
I tried to start with a very simple example in README and followed this thread https://github.com/jupyter/jupyter-js-services/issues/86

```
git clone https://github.com/jupyter/jupyter-js-services.git
cd jupyter-js-services
sudo npm install
```

Then I created another folder and install dependencies:
```
mkdir test
cd test
sudo npm install jupyter-js-services ws xmlhttprequest
```

I created index.ts with below content:

```
import {
  listRunningKernels, connectToKernel, startNewKernel, getKernelSpecs
} from 'jupyter-js-services';

// get a list of available kernels and connect to one
listRunningKernels({ baseUrl: 'http://localhost:8000' }).then(kernelModels => {
  var options = {
    baseUrl: 'http://localhost:8000',
    wsUrl: 'ws://localhost:8000',
    name: kernelModels[0].name
  }
  connectToKernel(kernelModels[0].id, options).then((kernel) => {
    console.log(kernel.name);
  });
});


// get info about the available kernels and start a new one
getKernelSpecs({ baseUrl: 'http://localhost:8888' }).then(kernelSpecs => {
  console.log('Default spec:', kernelSpecs.default);
  console.log('Available specs', Object.keys(kernelSpecs.kernelspecs));
  // use the default name
  var options = {
    baseUrl: 'http://localhost:8888',
    wsUrl: 'ws://localhost:8888',
    name: kernelSpecs.default
  }
  startNewKernel(options).then((kernel) => {
    // execute and handle replies
    var future = kernel.execute({ code: 'a = 1' } );
    future.onDone = () => {
      console.log('Future is fulfilled');
    }
    future.onIOPub = (msg) => {
      console.log(msg.content);  // rich output data
    }

    // restart the kernel and then send an inspect message
    kernel.restart().then(() => {
      var request = { code: 'hello', cursor_pos: 4, detail_level: 0};
      kernel.inspect(request).then((reply) => {
        console.log(reply.data);
      });
    });

    // interrupt the kernel and then send a complete message
    kernel.interrupt().then(() => {
      kernel.complete({ code: 'impor', cursor_pos: 4 } ).then((reply) => {
        console.log(reply.matches);
      });
    });

    // register a callback for when the kernel changes state
    kernel.statusChanged.connect((status) => {
      console.log('status', status);
    });

    // kill the kernel
    kernel.shutdown().then(() => {
      console.log('Kernel shut down');
    });
  });
});
```

Then I hit:
```
tsc index.ts
```

And it gave me this errors:

```
node_modules/jupyter-js-services/lib/config.d.ts(21,27): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/config.d.ts(32,111): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/config.d.ts(56,35): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(128,49): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(139,56): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(147,27): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(158,44): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(169,37): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(180,40): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(188,33): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(197,37): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(206,36): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(216,60): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(226,59): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(266,49): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(280,57): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(291,27): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(305,44): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(321,47): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(337,44): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(349,33): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(361,37): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(373,36): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(386,60): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/contents.d.ts(399,59): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(505,18): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(521,16): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(536,17): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(546,19): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(556,43): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(566,41): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(588,47): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(598,43): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(631,22): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(644,41): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(648,44): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(652,41): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/ikernel.d.ts(656,54): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/isession.d.ts(65,42): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/isession.d.ts(66,45): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/isession.d.ts(70,41): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/isession.d.ts(74,55): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/isession.d.ts(121,65): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/isession.d.ts(129,45): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/kernel.d.ts(17,41): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/kernel.d.ts(23,44): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/kernel.d.ts(29,41): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/kernel.d.ts(35,54): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/kernel.d.ts(48,66): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/kernel.d.ts(57,70): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/kernel.d.ts(68,66): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/kernel.d.ts(85,80): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/session.d.ts(18,42): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/session.d.ts(24,45): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/session.d.ts(31,41): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/session.d.ts(37,55): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/session.d.ts(52,72): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/session.d.ts(65,68): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-services/lib/session.d.ts(82,82): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-utils/lib/index.d.ts(130,76): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-utils/lib/index.d.ts(140,5): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-utils/lib/index.d.ts(152,14): error TS2304: Cannot find name 'Promise'.
node_modules/jupyter-js-utils/lib/index.d.ts(156,25): error TS2304: Cannot find name 'Thenable'.
```

If I use `node`, I got this:

```
# node index.js
API request failed (undefined):
# node index.ts
/Users/myusername/Documents/test/index.ts:3
import {
^^^^^^

SyntaxError: Unexpected token import
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:387:25)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:141:18)
    at node.js:933:3
```

So at this point I don't even know **how to get started**. Can you help?

**UPDATE 1**

Look like my `sudo npm run build` had error and I forgot to mention:

```
> jupyter-js-services@0.5.4 build /Users/myusername/Documents/jupyter-js-services
> npm run build:src && npm run build:test


> jupyter-js-services@0.5.4 build:src /Users/myusername/Documents/jupyter-js-services
> tsc --project src

src/session.ts(241,5): error TS7027: Unreachable code detected.

npm ERR! Darwin 14.3.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "build:src"
npm ERR! node v5.7.0
npm ERR! npm  v3.6.0
npm ERR! code ELIFECYCLE
npm ERR! jupyter-js-services@0.5.4 build:src: `tsc --project src`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the jupyter-js-services@0.5.4 build:src script 'tsc --project src'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the jupyter-js-services package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     tsc --project src
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs jupyter-js-services
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls jupyter-js-services
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/myusername/Documents/jupyter-js-services/npm-debug.log

npm ERR! Darwin 14.3.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "build"
npm ERR! node v5.7.0
npm ERR! npm  v3.6.0
npm ERR! code ELIFECYCLE
npm ERR! jupyter-js-services@0.5.4 build: `npm run build:src && npm run build:test`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the jupyter-js-services@0.5.4 build script 'npm run build:src && npm run build:test'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the jupyter-js-services package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run build:src && npm run build:test
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs jupyter-js-services
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls jupyter-js-services
npm ERR! There is likely additional logging output above.
```

Here is what in the `npm-debug.log`

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/bin/node', '/usr/local/bin/npm', 'run', 'build' ]
2 info using npm@3.6.0
3 info using node@v5.7.0
4 verbose run-script [ 'prebuild', 'build', 'postbuild' ]
5 info lifecycle jupyter-js-services@0.5.4~prebuild: jupyter-js-services@0.5.4
6 silly lifecycle jupyter-js-services@0.5.4~prebuild: no script for prebuild, continuing
7 info lifecycle jupyter-js-services@0.5.4~build: jupyter-js-services@0.5.4
8 verbose lifecycle jupyter-js-services@0.5.4~build: unsafe-perm in lifecycle true
9 verbose lifecycle jupyter-js-services@0.5.4~build: PATH: /usr/local/lib/node_modules/npm/bin/node-gyp-bin:/Users/myusername/Documents/jupyter-js-services/node_modules/.bin:/Users/myusername/.rvm/gems/ruby-2.2.1/bin:/Users/myusername/.rvm/gems/ruby-2.2.1@global/bin:/Users/myusername/.rvm/rubies/ruby-2.2.1/bin:/Users/myusername/Library/Android/sdk/tools:/Users/myusername/Library/Android/sdk/platform-tools:/opt/jpeg-9a:/opt/apache-maven-3.3.3/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/myusername/.rvm/bin
10 verbose lifecycle jupyter-js-services@0.5.4~build: CWD: /Users/myusername/Documents/jupyter-js-services
11 silly lifecycle jupyter-js-services@0.5.4~build: Args: [ '-c', 'npm run build:src && npm run build:test' ]
12 silly lifecycle jupyter-js-services@0.5.4~build: Returned: code: 1  signal: null
13 info lifecycle jupyter-js-services@0.5.4~build: Failed to exec build script
14 verbose stack Error: jupyter-js-services@0.5.4 build: `npm run build:src && npm run build:test`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/lifecycle.js:232:16)
14 verbose stack     at emitTwo (events.js:100:13)
14 verbose stack     at EventEmitter.emit (events.js:185:7)
14 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/spawn.js:24:14)
14 verbose stack     at emitTwo (events.js:100:13)
14 verbose stack     at ChildProcess.emit (events.js:185:7)
14 verbose stack     at maybeClose (internal/child_process.js:827:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
15 verbose pkgid jupyter-js-services@0.5.4
16 verbose cwd /Users/myusername/Documents/jupyter-js-services
17 error Darwin 14.3.0
18 error argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "build"
19 error node v5.7.0
20 error npm  v3.6.0
21 error code ELIFECYCLE
22 error jupyter-js-services@0.5.4 build: `npm run build:src && npm run build:test`
22 error Exit status 1
23 error Failed at the jupyter-js-services@0.5.4 build script 'npm run build:src && npm run build:test'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the jupyter-js-services package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     npm run build:src && npm run build:test
23 error You can get information on how to open an issue for this project with:
23 error     npm bugs jupyter-js-services
23 error Or if that isn't available, you can get their info via:
23 error     npm owner ls jupyter-js-services
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

Here is the versions I use:

```
# node --version
v5.7.0
# npm --version
3.6.0
# tsc --version
Version 1.8.7
```

**Comment by blink1073** [comment]
Hi @quanghoc, if you use the library from TypeScript, you'll need to include one or more of the typings files in https://github.com/jupyter/jupyter-js-services/tree/master/typings.

If you use the library from node, you'll have to change the imports from ES6 -> ES5 syntax, since node does not yet support ES6.

**Comment by quanghoc** [comment]
I have tried few more times and none of the below work.

**1. Add typings**

I copied `typings` folder and added below to my original `index.ts` top:

```
/// <reference path="./typings/es6-collections/es6-collections.d.ts" />
/// <reference path="./typings/es6-promise/es6-promise.d.ts" />
/// <reference path="./typings/expect.js/expect.js.d.ts" />
/// <reference path="./typings/mocha/mocha.d.ts" />
/// <reference path="./typings/requirejs/r.d.ts" />
/// <reference path="./typings/text-encoding/text-encoding.d.ts" />
/// <reference path="./typings/ws/ws.d.ts" />
/// <reference path="./typings/xmlhttprequest/xmlhttprequest.d.ts" />
```

Then I ran `tsc index.ts` (which no longer gave errors) and then `node index.js`. But I now got this error:

```
API request failed (undefined):
```

I went to `/node_modules/jupyter-js-services/lib/kernel.js` and found out that the original example in the page used port `8000` which is incorrect. The default port is `8888`. But even when I fixed that, it's still the same error.

**2. Cross check in Postman**

Well I was thinking to directly call the API myself to a running notebook (using `python -m notebook --NotebookApp.allow_origin="*"` as instructed to avoid cross-site issue). And I got the latest Notebook from https://github.com/jupyter/notebook. But then Postman ended up gave the blank return `[]` as well. 

<img width="959" alt="screen shot 2016-03-08 at 10 37 48 pm" src="https://cloud.githubusercontent.com/assets/2769972/13627455/5861d09c-e57f-11e5-8d2e-b280402b38f3.png">


So at this point, I think calling GET `http://localhost:8888/api/kernels` is wrong? That's where I left off.

**Any help here?**

**Comment by blink1073** [comment]
Hi @quanghoc, can you `console.log` the `url` and the `ajaxSettings` that are being used for you in that function?

**Comment by quanghoc** [comment]
Yes I did and it showed what expected:

```
http://localhost:8888/api/kernelspecs
{ method: 'GET', dataType: 'json' }
```

**Comment by blink1073** [comment]
Can you debug the `utils.ajaxRequest` call here https://github.com/jupyter/jupyter-js-utils/blob/master/src/index.ts#L268, and print the `req.response`, `req.status`, and `err.message`?

**Comment by quanghoc** [comment]
It showed:

```
[ReferenceError: XMLHttpRequest is not defined]
```

But when I added below to index.ts, I got `tsc index.ts` error (which was why I didn't included it in beginning):

```
import { XMLHttpRequest } from "xmlhttprequest";
import { default as WebSocket } from 'ws';

global.XMLHttpRequest = XMLHttpRequest;
global.WebSocket = WebSocket;
```

**Comment by blink1073** [comment]
Does it work if you add this before the `global` lines? 

```typescript
// stubs for node global variables
declare var global: any;
```

**Comment by quanghoc** [comment]
That solved the `global` var error but still got the `ws` error.

After I did some google, I had to use this:

```
import WebSocket = require('ws');
```

It printed out the kernel specs as expected. But also did the following which I am not sure if it is supposed to do that 

```
tarting WebSocket: ws://localhost:8888/api/kernels/9fa9df4c-763d-4b90-9e55-c237fe0e8671
Kernel: restarting (9fa9df4c-763d-4b90-9e55-c237fe0e8671)
status Promise { <pending> }
Kernel:  (9fa9df4c-763d-4b90-9e55-c237fe0e8671)
status Promise { <rejected> [Error: Kernel is not ready to send a message] }
Connection lost, reconnecting in 1 seconds.
Kernel shut down
Starting WebSocket: ws://localhost:8888/api/kernels/9fa9df4c-763d-4b90-9e55-c237fe0e8671
```

**Comment by blink1073** [comment]
That could be, if the kernel was not ready the first time the request was made.

**Comment by blink1073** [comment]
Can you summarize what you had to do to get this working?  Even better, a PR with some additions to the README would be most welcome :smile:.

**Comment by KimStebel** [comment]
I can't get this to work in the browser either. I've followed all the hints in this ticket, but I'm not getting anywhere. Can you please add a full working example of how to use the library? All of this seems needlessly complicated.

**Comment by blink1073** [comment]
Message received: WIP https://github.com/jupyter/jupyter-js-services/pull/111

**Comment by quanghoc** [comment]
I was meant to provide the example folder but then got stuck in figuring
out how to do cell execution... :P
